src/main/groovy/edu/oregonstate/mist/musicapi/resources/AlbumResource.groovy

```diff
@@ -0,0 +1,268 @@
+import javax.ws.rs.PathParam
+import javax.ws.rs.Produces
+import javax.ws.rs.Consumes
+import javax.ws.rs.core.Response
+import javax.ws.rs.core.Response.ResponseBuilder
+import org.skife.jdbi.v2.DBI
+import org.skife.jdbi.v2.Handle
+import org.skife.jdbi.v2.util.IntegerMapper
+
+@Path('/album')
+@groovy.transform.TypeChecked
+class AlbumResource extends Resource {
+    private DBI dbi
+    private Pattern releaseDatePattern = Pattern.compile(/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/) // YYYY-MM-DD
+
+    public AlbumResource(DBI dbi) {
+        this.dbi = dbi
+    }
+
+    /**
+     * Responds to GET requests by returning an album
+     *
+     * @return album
+     */
+    @GET
+    @Path('{id}')
+    @Produces("application/json")
+    Response getAlbum(@Auth AuthenticatedUser authenticatedUser, @PathParam("id") int id) {
+        def h = this.dbi.open()
+        try {
+            def result = this.getAlbumById(h, id)
+            if (result == null) {
+                return this.notFound().build()
+            }
+            return this.ok(result).build()
+        } finally {
+            h.close()
+        }
+    }
+
+    /**
+     * Creates a new album
+     */
+    @POST
+    @Consumes("application/json")
+    @Produces("application/json")
+    Response createAlbum(@Auth AuthenticatedUser _, Album album) {
+        def h = this.dbi.open()
+        try {
+            // Get artist id
+            def artistId = this.getOrCreateArtist(h, album.artist)
+
+            // Get status id
+            def statusId = this.getStatus(h, album.status)
+            if (statusId == null) {
+                return this.badRequest('invalid status').build()
+            }
+
+            // validate release date
+            if (album.released != null) {
+                if (!releaseDatePattern.matcher(album.released).matches()) {
+                    return this.badRequest('invalid release date').build()
+                }
+            }
+
+            def q = h.createStatement('''
+                INSERT INTO mus_album (id, title, artist_id, edition, status, released, created)
+                VALUES (
+                    NULL, :title, :artist_id, :edition, :status_id,
+                    NULL, cast(SYSTIMESTAMP at time zone 'UTC' as date))
+            ''')
+            q.bind('title', album.title)
+            q.bind('artist_id', artistId.intValue())
+            q.bind('edition', album.edition)
+            q.bind('status_id', statusId.intValue())
+
+            def gk = q.executeAndReturnGeneratedKeys()
+            def rowid = gk.first().get('rowid')
+
+            // Set release date, if provided
+            if (album.released) {
+                q = h.createStatement('''
+                    UPDATE mus_album
+                    SET released = to_date(:released, 'YYYY-MM-DD')
+                    WHERE rowid = :rowid''')
+                q.bind('released', album.released)
+                q.bind('rowid', rowid)
+                q.execute()
```

Comment:
This works in this example and it's a learning app. One question for you though is what happens to the rowid when a record is deleted at the same time that you're adding a new record?